File: backend/src/main/java/com/chikecan/backend/entity/ExperienceLevel.java
package com.chikecan.backend.entity;

/**
 * 累計XPからレベル・レベル内XP・次のレベルまでのXPを計算する。
 * レベルや残りXPをDBへ重複保存せず、常に累計XP(experience)から算出するための
 * 唯一の計算箇所。UserエンティティとAppUserDetailsの両方から利用する。
 */
public final class ExperienceLevel {

  private static final int EXPERIENCE_PER_LEVEL = 100;

  private ExperienceLevel() {
  }

  public static int level(int experience) {
    return experience / EXPERIENCE_PER_LEVEL + 1;
  }

  public static int currentLevelExperience(int experience) {
    return experience % EXPERIENCE_PER_LEVEL;
  }

  public static int experienceToNextLevel(int experience) {
    return EXPERIENCE_PER_LEVEL - currentLevelExperience(experience);
  }

  public static int progressPercentage(int experience) {
    return currentLevelExperience(experience);
  }
}

File: frontend/src/utils/experienceLevel.ts
/**
 * 累計XPからレベル・レベル内XP・次のレベルまでのXPを計算する。
 * バックエンドのExperienceLevel(entity)と同じ式を用いる、フロント側の唯一の計算箇所。
 */
export const EXPERIENCE_PER_LEVEL = 100;

export function levelFromExperience(experience: number): number {
  return Math.floor(experience / EXPERIENCE_PER_LEVEL) + 1;
}

export function currentLevelExperience(experience: number): number {
  return experience % EXPERIENCE_PER_LEVEL;
}

export function experienceToNextLevel(experience: number): number {
  return EXPERIENCE_PER_LEVEL - currentLevelExperience(experience);
}

export function experienceProgressPercentage(experience: number): number {
  return currentLevelExperience(experience);
}
